### PR TITLE
Feature/weighted traversal

### PR DESCRIPTION
**Pull request by maierlars**
### Scope & Purpose

https://github.com/arangodb/enterprise/pull/510

PR adds a `mode` field to traversal options (deprecating old `bfs: true`) now allowing following value: `dfs` (default), `bfs` and `weighted`. In weighted mode there is now a `weightAttribute` and `defaultWeight` attribute with same semantics as for shortest path traversals. Paths are enumerate by increasing length. Furthermore returned path has a new attribute `weights` that contains the weights at each individual step.

For example:
```
for x, v, p in 0..10 outbound "places/York" graph "kShortestPathsGraph" options {order: "weighted", weightAttribute: "travelTime", uniqueVertices: "path"}
    FILTER p.edges[*].travelTime ALL < 3
    let totalTime = LAST(p.weights)
    FILTER totalTime < 6
    SORT totalTime DESC
    return {path: p.vertices[*]._key, weight: LAST(p.weights), weights: p.edges[*].travelTime}
```

returns
path | weight | weights
-- | -- | --
["York","London","Birmingham","Carlisle"] | 5.3 | [1.8,2.5,1]
["York","London","Birmingham"] | 4.3 | [1.8,2.5]
["York","London","Brussels"] | 4.3 | [1.8,2.5]
["York","London"] | 1.8 | [1.8]
["York"] | 0 | []

I'd like to ask for your feed back on this. Of course I will add tests later. It should work with all other options like `uniqueVertices` and `uniqueEdges`. Important case `uniqueVertices: "global"` <3

- [x] Strictly *new functionality* (i.e. a new feature / new option, no need for porting)
- [x] The behavior in this PR can be (and was) *manually tested* (support / qa / customers can test it)
- [x] The behavior change can be verified via automatic tests

### Testing & Verification

This PR adds tests that were used to verify all changes:

- [ ] Added new **integration tests** (i.e. in shell_server / shell_server_aql)

Additionally:
- [x] I ensured this code runs with ASan / TSan or other static verification tools

### Documentation

Will add if you like the change.

- [x] Added a *Changelog Entry* (referencing the corresponding public or internal issue number)
- [x] Docs PR: https://github.com/arangodb/docs/pull/558

### CODE REVIEW

**Comment by Simran-B** [comment]
> Paths are enumerate by increasing length

That makes `mode: "weighted"` breadth-first, right?

**Comment by maierlars** [comment]
> > Paths are enumerate by increasing length
> 
> That makes `mode: "weighted"` breadth-first, right?

Could you elaborate on that? There is now "bfs", "dfs" and "weighted". "weighted" without weights is a _slower_ form of "bfs". Did that answer your question?

**Comment by maierlars** [comment]
https://jenkins.arangodb.biz/job/arangodb-matrix-pr/11985/

**Comment by maierlars** [comment]
https://jenkins.arangodb.biz/job/arangodb-matrix-pr/11995/

**Comment by maierlars** [comment]
https://jenkins.arangodb.biz/job/arangodb-matrix-pr/11996/

**Comment by maierlars** [comment]
https://jenkins.arangodb.biz/job/arangodb-matrix-pr/11998/

**Comment by maierlars** [comment]
https://jenkins.arangodb.biz/job/arangodb-matrix-pr/11999/

**Comment by maierlars** [comment]
image

File: lib/Basics/VelocyPackHelper.h [hunk 1]
```diff
@@ -88,6 +89,42 @@ struct VPackHashedSlice {
   ~VPackHashedSlice() = default;
 };
 
+namespace detail {
+template <class>
+inline constexpr bool always_false_v = false;
+template <typename... Ts, std::size_t... Is>
+auto unpackTuple(VPackArrayIterator& iter, std::index_sequence<Is...>) {
```
Comment (mchacki) [inline lib/Basics/VelocyPackHelper.h:96]: I like this, very helpful!

File: tests/Aql/TraversalExecutorTest.cpp [hunk 2]
```diff
@@ -217,6 +217,11 @@ class TraverserHelper : public Traverser {
     return false;
   }
 
+  bool getVertex(arangodb::velocypack::StringRef vertex, size_t depth) override {
+    // Implement
```
Comment (mchacki) [inline tests/Aql/TraversalExecutorTest.cpp:221]:
```suggestion
```

I think this is never needed in the test, just to please the API

File: tests/js/server/aql/aql-graph-weighted-traversal.js [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+const jsunity = require("jsunity");
+const db = require("@arangodb").db;
+const gm = require("@arangodb/general-graph");
+const _ = require("underscore");
+
+/*
+
+  1 - 1 > 2 - 10 > 3 -> 5
+   \       /
+     4 -> 6
```
Comment (mchacki) [inline tests/js/server/aql/aql-graph-weighted-traversal.js:13]: Thi is not aligned with data.
6 -> 3 uses wight 0.5 which is missing here.

**Review by mchacki — CHANGES_REQUESTED** [review]
Just one minor item in the edge cursor.
Which checks 1 depth too much.
Everything else I like very much good job :+1:

**Comment by maierlars** [comment]
https://jenkins.arangodb.biz/job/arangodb-matrix-pr/12023/

**Comment by maierlars** [comment]
https://jenkins.arangodb.biz/job/arangodb-matrix-pr/12032/

**Review by maierlars — COMMENTED** [review]
foobar

**Comment by maierlars** [comment]
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr/12068/

**Comment by maierlars** [comment]
Hello world!

**Comment by maierlars** [comment]
Hello world!

**Comment by maierlars** [comment]
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr/12069/

**Comment by maierlars** [comment]
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr/12070/

**Comment by maierlars** [comment]
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr/12076/

**Review by mchacki — APPROVED** [review]
LGTM

**Pull request merged**